scvi/external/sysvi/_module.py

```diff
@@ -0,0 +1,445 @@
+            losses = self.loss(
+                tensors=tensors,
+                inference_outputs=inference_outputs_merged,
+                generative_outputs=generative_outputs_merged,
+                **loss_kwargs,
+            )
+            return inference_outputs_merged, generative_outputs_merged, losses
+        else:
+            return inference_outputs_merged, generative_outputs_merged
+
+    def loss(
+        self,
+        tensors,
+        inference_outputs,
+        generative_outputs,
+        kl_weight: float = 1.0,
+        reconstruction_weight: float = 1.0,
+        z_distance_cycle_weight: float = 2.0,
+    ):
+        x_true = tensors[REGISTRY_KEYS.X_KEY]
+
+        # Reconstruction loss
+
+        def reconst_loss_part(x_m, x, x_v):
+            """Compute reconstruction loss"""
+            return torch.nn.GaussianNLLLoss(reduction="none")(x_m, x, x_v).sum(dim=1)
+
+        # Reconstruction loss
+        reconst_loss_x = reconst_loss_part(
+            x_m=generative_outputs["x_m"], x=x_true, x_v=generative_outputs["x_v"]
+        )
+        reconst_loss = reconst_loss_x
+
+        # Kl divergence on latent space
+        kl_divergence_z = self.prior.kl(
+            m_q=inference_outputs["z_m"],
+            v_q=inference_outputs["z_v"],
+            z=inference_outputs["z"],
+        )
+
+        def z_dist(z_x_m: torch.Tensor, z_y_m: torch.Tensor):
+            """MSE loss between standardised inputs with standardizer fitted on concatenation of both inputs"""
+            # Standardise data (jointly both z-s) before MSE calculation
+            z = torch.concat([z_x_m, z_y_m])
+            means = z.mean(dim=0, keepdim=True)
+            stds = z.std(dim=0, keepdim=True)
+
+            def standardize(x):
+                return (x - means) / stds
+
+            return torch.nn.MSELoss(reduction="none")(standardize(z_x_m), standardize(z_y_m)).sum(
+                dim=1
+            )
+
+        z_distance_cyc = z_dist(z_x_m=inference_outputs["z_m"], z_y_m=inference_outputs["z_cyc_m"])
+
+        loss = (
+            reconst_loss * reconstruction_weight
+            + kl_divergence_z * kl_weight
+            + z_distance_cyc * z_distance_cycle_weight
+        )
+
+        return LossOutput(
+            n_obs_minibatch=loss.shape[0],
+            loss=loss.mean(),
+            extra_metrics={
+                "reconstruction_loss": reconst_loss.mean(),
+                "kl_local": kl_divergence_z.mean(),
+                "z_distance_cycle": z_distance_cyc.mean(),
+            },
+        )
+
+    @staticmethod
+    def random_select_systems(system: torch.Tensor) -> torch.Tensor:
+        """For every cell randomly selects a new system that is different from the original system
+
+        Parameters
+        ----------
+        system
+            One hot encoded system information for each cell
+
+        Returns
+        -------
+        One hot encoding of newly selected system for each cell
+
+        """
```

Comment:
Requires some thinking to rewrite this function. It's currently quite clunky and likely slow for the performed task.